File: src/model2.jl
include("OhMyJulia.jl")
include("Fire.jl")
include("Keras.jl")

using OhMyJulia
using Fire
using Keras
using StatsBase
using HDF5

const model = let
    input1 = Keras.Input(shape=(256, 63, 10))
    input2 = Keras.Input(shape=(256,  1, 10))
    input3 = Keras.Input(shape=(2,))

    conv = input1 |>
        Keras.Convolution2D(16, (1, 3), activation="relu", padding="same") |>
        Keras.MaxPooling2D((1, 3), strides=(1, 3)) |>
        Keras.Convolution2D(32, (1, 5), activation="relu", padding="same") |>
        Keras.Convolution2D(32, (1, 5), activation="relu", padding="same") |>
        Keras.MaxPooling2D((1, 3), strides=(1, 3)) |>
        Keras.Convolution2D(32, (1, 7), activation="relu", padding="valid")

    fc = input1 |>
        Keras.Convolution2D(32, (1, 63), activation="relu", padding="valid")

    res = [conv, fc, input2] |>
        Keras.Concatenate(3) |>
        Keras.Convolution2D(64, (1, 1), activation="relu", padding="same") |>
        Keras.AveragePooling2D((256, 1), strides=(256, 1)) |>
        Keras.Flatten()

    output = [res, input3] |>
        Keras.Concatenate(1) |>
        Keras.Dense(32, activation="sigmoid") |>
        Keras.Dense(1, activation="sigmoid")

    Keras.Model(inputs=[input1, input2, input3], outputs=[output])
end

const callbacks = [Keras.ModelCheckpoint("model2_weight.{epoch:02d}-{val_acc:.4f}.h5", monitor="val_acc")]

const phase_one = readdir(".") ~ filter(x->startswith(x, "model2_weight.14"))

function prepare_data()
    images  = readdir(".") ~ filter(x->endswith(x, ".image")) ~ map(i"1:end-6")
    txts    = readdir(".") ~ filter(x->endswith(x, ".txt"))   ~ map(i"1:end-4")
    samples = intersect(images, txts)

    results = map(samples) do sam
        image = open(sam * ".image")
        txt   = open(sam * ".txt")
        txt   = map(split, readlines(txt))
        y     = map(x->parse(f32, x[4]), txt)
        freq  = map(x->parse(f32, x[5]), txt)
        depth = map(x->parse(i32, x[6]), txt)
        X     = Array{f32}(length(y), 256, 64, 10)
        for (i, d) in enumerate(depth)
            reads = [read(image, f32, 64, 10) for i in 1:d]
            if d > 256
                for (j, r) in enumerate(sample(reads, 256, replace=false, ordered=true))
                    X[i, j, :, :] = r
                end
            else
                for (j, r) in enumerate(reads)
                    X[i, j, :, :] = r
                end
                X[i, d+1:end] = 0.
            end
        end
        X[:, :, 1:63, :], X[:, :, [64], :], [freq min.(depth, 2048)./2048], y
    end

    X1 = [map(i"1", results)...;]
    X2 = [map(i"2", results)...;]
    X3 = [map(i"3", results)...;]
    y  = [map(i"4", results)...;]

    h5open("model2_data.h5", "w") do f
        write(f, "X1", X1)
        write(f, "X2", X2)
        write(f, "X3", X3)
        write(f,  "y",  y)
    end

    X1, X2, X3, y
end

@main function train()
    if isfile("model2_data.h5")
        X1, X2, X3, y = h5open("model2_data.h5") do f
            read(f, "X1"), read(f, "X2"), read(f, "X3"), read(f, "y")
        end
    else
        prt(STDERR, now(), "preparing data")
        X1, X2, X3, y = prepare_data()
    end

    prt(STDERR, now(), "start training")
    if !isempty(phase_one)
        model[:compile](Keras.SGD(lr=1e-3, decay=1e-4), "binary_crossentropy", metrics=["accuracy"])
        model[:load_weights](phase_one[])
        model[:fit]([X1, X2, X3], y, batch_size=256, epochs=20, validation_split=.02, callbacks=callbacks, initial_epoch=15)
    else
        model[:compile](Keras.SGD(lr=2e-3, momentum=.95), "binary_crossentropy", metrics=["accuracy"])
        model[:fit]([X1, X2, X3], y, batch_size=64, epochs=15, validation_split=.02, callbacks=callbacks)
    end
end

File: model2.py
#!/usr/bin/env python

"model2: convolution only across bases"

import os
import keras
import random
import numpy as np

def get_model():
    input1 = keras.layers.Input(shape=(256, 63, 10))
    input2 = keras.layers.Input(shape=(256,  1, 10))
    input3 = keras.layers.Input(shape=(2,))

    x = keras.layers.Convolution2D(16, (1, 3), activation="relu", padding="same")(input1)
    x = keras.layers.MaxPooling2D((1, 3), strides=(1, 3))(x)
    x = keras.layers.Convolution2D(32, (1, 5), activation="relu", padding="same")(x)
    x = keras.layers.Convolution2D(32, (1, 5), activation="relu", padding="same")(x)
    x = keras.layers.MaxPooling2D((1, 3), strides=(1, 3))(x)
    conv = keras.layers.Convolution2D(32, (1, 7), activation="relu", padding="valid")(x)

    fc = keras.layers.Convolution2D(32, (1, 63), activation="relu", padding="valid")(input1)

    x = keras.layers.Concatenate(3)([conv, fc, input2])
    x = keras.layers.Convolution2D(64, (1, 1), activation="relu", padding="same")(x)
    x = keras.layers.AveragePooling2D((256, 1), strides=(256, 1))(x)
    res = keras.layers.Flatten()(x)

    x = keras.layers.Concatenate(1)([res, input3])
    x = keras.layers.Dense(32, activation="sigmoid")(x)
    x = keras.layers.Dense(1, activation="sigmoid")(x)

    model = keras.models.Model(inputs=[input1, input2, input3], outputs=[x])
    model.compile(keras.optimizers.SGD(lr=1e-3, momentum=.95), "binary_crossentropy", metrics=["accuracy"])
    return model

def collect_data():
    files = os.listdir('.')
    images = [x[:-3] for x in files if x.endswith('.image')]
    samples = [x for x in images if x+".txt" in files]

    data = []
    for sam in samples:
        with open(sam+".image") as image, open(sam+".txt") as txt:
            for line in txt:
                line = line.split()
                y, freq, depth = int(line[3]), float(line[4]), int(line[5])
                X = np.zeros(dtype=float, shape=(256,64,10))
                reads = [np.fromfile(image, dtype=float, count=64*10)]
                if depth > 256:
                    reads = [reads[x] for x in sorted(random.sample(range(depth), count=256))]
                for i in range(len(reads)):
                    X[i, :, :] = read
                data.append((X[:, 0:63, :], X[:, [63], :], np.array([freq, min(depth, 2048)/depth]), y))

    with open("model2.data", "w") as out:
        np.array([len(data)]).tofile(out)
        np.array([d[3] for d in data]).tofile(out)
        for X, _, _, _ in data:
            X.tofile(out)
        for _, X, _, _ in data:
            X.tofile(out)
        for _, _, X, _ in data:
            X.tofile(out)

def train():
    with open("model1.data") as f:
        n = np.fromfile(f, dtype=int, count=1)[0]
        y = np.fromfile(f, dtype=int, count=n)
        X1 = np.fromfile(f, dtype=float, count=n*256*63*10).reshape(n, 256, 63, 10)
        X2 = np.fromfile(f, dtype=float, count=n*256*10).reshape(n, 256, 1, 10)
        X3 = np.fromfile(f, dtype=float, count=n*2).reshape(n, 2)

    model = get_model()
    callbacks = [keras.callbacks.ModelCheckpoint("model2_weight.{epoch:02d}-{val_acc:.4f}.h5", monitor="val_acc")]
    
    model.fit(X, y, batch_size=64, epochs=20, validation_split=.02, callbacks=callbacks)

if __name__ == '__main__':
    import fire
    fire.Fire({ 'train': train, 'collect_data': collect_data })